### ISSUE TITLE
Appending of pending rewards is conflicting for large networks

### ISSUE THREAD

**Issue by nzpr** [rchain/rchain#3349]
## Overview
When running tests on reasonably large number of validators (30+), conflict detection between some blocks can lead to false positive.
Conflict that is detected is race for base produce - it is destroyed in both event logs. Thought it should not. At least it is expected for conflict resolution outcome to be always the same. This happens in precharge or refund, it is not clear where exactly. But once blocks of two validators are considered conflicting, the same happens on each subsequent attempt to merge newly created blocks. Channel that is conflicting is always the same.

### Impact
-- not clear

### Dependencies
--

### Design
--

### Steps to implement the solution
Best solution is to have debugger for Rholang, so `term - AST - ADT - event log hashes` translation is visible. But intuition says that the bug might be related to peek (also see https://github.com/rchain/rchain/issues/3200, https://github.com/rchain/rchain/issues/3268). Destroyed produce means linear consume happened on the same produce both in main and merging branches. And prechrge + refund should use peeks on potentially conflicting produces. 

### Steps for integration testing
--

### Resource estimation
--

**Comment by zsluedem** [comment]
https://github.com/rchain/rchain/runs/2490010239?check_suite_focus=true
This can be reproduced by test. But it is more difficult with less nodes.

**Comment by nzpr** [comment]
I cannot reproduce this with new merging logic in `deploy-merge` feature branch 59f787d050219213ef2d8cd398dd87ebe1a1eff8 so closing this issue

**Comment by zsluedem** [comment]
https://github.com/rchain/rchain/runs/2801580638?check_suite_focus=true
reopen this issue because of the action above

The test action is based on dev. Maybe the bug is fixed in deploymerge branch. @nzpr

**Comment by nzpr** [comment]
@zsluedem yes, dev branch has initial implementation of block merge, which has this issue. It should not be present in `feature/deploymerge`

**Comment by nzpr** [comment]
After this PR https://github.com/rchain/rchain/pull/3444#discussion_r654188163 the bug is observable again

**Comment by nzpr** [comment]
Here is some more data for network of 50 validators:
49 blocks are merged into base state, so 49 branches are present and conflicts against combination of then is computed. 

Here is size of conflicts for each branch
```
conflict map values size: 1: 1: 0: 0: 2: 0: 0: 1: 0: 0: 0: 0: 1: 0: 1: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 2: 0: 0: 0: 0: 0: 0: 0: 1: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 2: 0: 0
```
As it is shown, some branches conflict with 2 another, some with only one, some does not have conflict with any.

**Comment by nzpr** [comment]
Here is some more data for network of 50 validators:

49 blocks are merged into base state, so 49 branches are present and conflicts against combination of then is computed. 
Here is size of conflicts with others for each branch
```
1: 1: 0: 0: 2: 0: 0: 1: 0: 0: 0: 0: 1: 0: 1: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 2: 0: 0: 0: 0: 0: 0: 0: 1: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 0: 2: 0: 0
```
As it is shown, some branches conflict with 2 another, some with only one, some does not have conflict with any.

**Comment by nzpr** [comment]
Conflict rule that is triggered is always the same - race for the same event, particularly for the same Produce.
Given two event logs, produces consumed are calculated in the following way
```
producesConsumedRef.update(s => s ++ c.produces).whenA(c.peeks.isEmpty)
```
so when COMM event is found and its `peeks` field is empty - all produces are recorded as consumed.
The problem that two event logs have the same produce recorded in this set.

**Comment by nzpr** [comment]
Looks like it is some problem with how peeks are exposed in COMM events. Here is example of conflict found.
```
Race for produce Produce(channels: Blake(d2d08f58fa9f0c6d1dea1bf40e392d6d24da4032337226cc6b24a62e94a6864a), hash: Blake(3e9f7ee648d4b691e0704c83864c653fcce4306a78b4a72db99a31adbd51d74d)) found
exists in pre state of A: true
exists in pre state of B: true
has linear produce record in A: true
has persistent produce record in A: false
has linear produce record in B: true
has persistent produce record in B: false
```
Which means that some linear produce exists in base state, has been is consumed and the same produce created. Which means it is peek.
But `peeks` field is empty in COMM event.

**Comment by nzpr** [comment]
Worth mentioning - that this peek log behaviour is non deterministic, sometimes it misses peek in COMM log field.

**Comment by nzpr** [comment]
After talking with @tgrospic intuition developed that this is a legit conflict case due to some buggy/wrong rholang implementation.
Further investigations led to this: for test (https://github.com/rchain/rchain/pull/3445)

the following conflicts map is calculated 
CONFLICT MAP (block/block): 
```
- - - - - - - - C - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - C - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - C - - - - - -
C - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - C
- C - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - C - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - C - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - C - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - C - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - C - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
- - - - - - - - - - - - - - - - - - - - - - - - C - - - - - - - - - - - - - - - - - - - - - - - -
```
6 block pairs are conflicting
 
This is data on conflicts
```
Race for produce Produce(channels: Blake(bfa9446540eac8d0794bd7073d9b11294bb6830b18649514c4279a506932040e), hash: Blake(9254a4e9560e288fd4fbc6d9c2f6fa938d596ed27903d473622e0d4dcbcf863c)) found, existsInPreStateA: true
existsInPreStateB: true
has linear produce record in A: true
has persistent produce record in A: false
has linear produce record in B: true
has persistent produce record in B: false
has peek record in A: true
has consume produce record in A: true
has peek record in B: true
has consume produce record in B: true
produce occurrences in a: 2
produce occurrences in b: 2

racesForSameIOEvent: List(Blake(bfa9446540eac8d0794bd7073d9b11294bb6830b18649514c4279a506932040e))
Race for produce Produce(channels: Blake(bf5a55a4d7213f7782c8324d5636d0fb02adc308bad6b78262d00530fcad35a5), hash: Blake(d98c08433b187c48d7303bd1d23cc79f0d3720af233f94d7653762c2592f25b0)) found, existsInPreStateA: true
existsInPreStateB: true
has linear produce record in A: true
has persistent produce record in A: false
has linear produce record in B: true
has persistent produce record in B: false
has peek record in A: true
has consume produce record in A: true
has peek record in B: true
has consume produce record in B: true
produce occurrences in a: 2
produce occurrences in b: 2

racesForSameIOEvent: List(Blake(bf5a55a4d7213f7782c8324d5636d0fb02adc308bad6b78262d00530fcad35a5))
Race for produce Produce(channels: Blake(ef4f7e9bbaa79b4bf7ba1a7282e8ed438be6a11514465e08387a19a9506dcc6d), hash: Blake(f73d7f364fa9656a7ee63f118f0736653a4807ab319ba80aef0e9dbe2951c031)) found, existsInPreStateA: true
existsInPreStateB: true
has linear produce record in A: true
has persistent produce record in A: false
has linear produce record in B: true
has persistent produce record in B: false
has peek record in A: true
has consume produce record in A: true
has peek record in B: true
has consume produce record in B: true
produce occurrences in a: 2
produce occurrences in b: 2

racesForSameIOEvent: List(Blake(ef4f7e9bbaa79b4bf7ba1a7282e8ed438be6a11514465e08387a19a9506dcc6d))
Race for produce Produce(channels: Blake(898d698ef359472a3b474eb6d8b3bc0ac88b0462f4d8ceb3625565bb2166954a), hash: Blake(75097b991d25d9c17f500d6ccace64ee361bc7b4dab770b94a27e8335ef4aed3)) found, existsInPreStateA: true
existsInPreStateB: true
has linear produce record in A: true
has persistent produce record in A: false
has linear produce record in B: true
has persistent produce record in B: false
has peek record in A: true
has consume produce record in A: true
has peek record in B: true
has consume produce record in B: true
produce occurrences in a: 2
produce occurrences in b: 2

racesForSameIOEvent: List(Blake(898d698ef359472a3b474eb6d8b3bc0ac88b0462f4d8ceb3625565bb2166954a))
Race for produce Produce(channels: Blake(2a453b217b8c4e43e539241de7e74e2c9da44b86691469936fb5a66c3c34cc71), hash: Blake(0d94469a9cecfdde6e303403be7b70ea1996e3d9d79083a3df17dcd7434e0147)) found, existsInPreStateA: true
existsInPreStateB: true
has linear produce record in A: true
has persistent produce record in A: false
has linear produce record in B: true
has persistent produce record in B: false
has peek record in A: true
has consume produce record in A: true
has peek record in B: true
has consume produce record in B: true
produce occurrences in a: 2
produce occurrences in b: 2

racesForSameIOEvent: List(Blake(2a453b217b8c4e43e539241de7e74e2c9da44b86691469936fb5a66c3c34cc71))
Race for produce Produce(channels: Blake(019b440aaf735c3b6ee8c99815debff6e6646784d93f3c581eb0c079bb059669), hash: Blake(6c159632a8c3cbb9c25343d48fff3813f3afd5989ada5bf5007732bb6c5fd070)) found, existsInPreStateA: true
existsInPreStateB: true
has linear produce record in A: true
has persistent produce record in A: false
has linear produce record in B: true
has persistent produce record in B: false
has peek record in A: true
has consume produce record in A: true
has peek record in B: true
has consume produce record in B: true
produce occurrences in a: 2
produce occurrences in b: 2
```

As it can be seen, conflict frame is the same for all occurrences. It is a linear produce, which exists in preState of both blocks.
It is met in two COMM - one linear and one peek. Produce as IOEvent is present 2 times in event log, which might mean that one produce is generated by peek, and another one is updating of a value consumed.

Therefore, it looks like it's a real conflict and should be investigated.
As @zsluedem suggested, this might be this piece of code
https://github.com/rchain/rchain/blob/dev/casper/src/main/resources/Registry.rho#L157-L205
that produces conflicts non deterministically.

**Comment by tgrospic** [comment]
I pushed a commit to #3445 which prints channel @zsluedem suggested and it shows that conflicting channels are indeed TreeHashMap insert nodes.

[tree-hash-map-channels.log](https://github.com/rchain/rchain/files/6680337/tree-hash-map-channels.log)

**Comment by nzpr** [comment]
Final resolution for this issue is that
https://github.com/rchain/rchain/blob/44eb1194558d97e7dc4b16154af8414e6ff32525/casper/src/main/resources/PoS.rhox#L695 this produce is conflicting, because pending rewards are stored in TreeHashMap with depth = 1, which makes some names storing pending rewards for different pub keys - the same. Issue is renamed.

**Comment by tgrospic** [comment]
Closing this issue because new issue is created to track this fix for Hard Fork 1 #3461.
